src/afrolabs/components/kafka.clj

```diff
@@ -145,17 +145,29 @@
 
 (deftype KafkaProducingCompletionCallback [msg delivered-ch]
   Callback
-  (onCompletion [_ _ ex]
+  (onCompletion [_ meta-data ex]
     (trace (str "Firing onCompletion for msg. "))
     ;; TODO ex may contain non-retriable exceptions, which must be used to indicate this component is not healthy
     (when ex
-      (warn ex "Error while getting confirmation of producing record."))
+      (trace "Forwarding delivery exception...")
+      (csp/go
+        (csp/>!! delivered-ch ex)
+        (csp/close! delivered-ch)
+        (trace "when exception onCompletion done.")))
     (when-not ex
       (trace "Forwarding delivery notification...")
       (csp/go
-        (csp/>! delivered-ch msg)
+        (csp/>! delivered-ch (merge msg
+                                    (cond-> {:topic     (.topic ^RecordMetadata meta-data)
+                                             :partition (.partition ^RecordMetadata meta-data)}
+
+                                      (.hasOffset ^RecordMetadata meta-data)
+                                      (assoc :offset (.offset ^RecordMetadata meta-data))
+
+                                      (.hasTimestamp ^RecordMetadata meta-data)
+                                      (assoc :timestamp (t/instant (.timestamp ^RecordMetadata meta-data))))))
         (csp/close! delivered-ch)
-        (trace "onCompletion done.")))))
+        (trace "onCompletion delivered result.")))))
```

Comment:
We enhance the capabilities of the _produce_ record's `:delivered-ch` channel. This channel used to received failures or exceptions (a type of failure) or in the case of success, the original record with `{:topic ... :value ...}` data. 

After this change
- more exceptions (failures) will be sent into the channel, 
- and the channel will be closed. 
- There was a way for the channel to remain open before.
- New fields are now available `:partition` & `:offset`. 
- This allows for creation of _Consistency Token_ values.